alexstrat/photoroom-test-assignment
# feat: add foldering features

This PR implements these user stories:
- As a user, I can see the list of images previously sent to the server from my computer. They all appear in an "Untitled Folder" folder by default.
- As a user, I can create folders
- As a user, I can move images between folders. An image can only be in one folder at a time.
- As a a user, when I refresh the page, the folders and images are still showing

https://www.loom.com/share/5e86b6aff0fd4eacb1c3d6720ecf0748

## How it was implemented
- `useImageLibrary` takes care of interacting with library state. The persistence is achieved via LocalStorage (using `react-storage-hooks`)
- `<LibrarySideBar />` displays the a tree view of folder and image and have navigation and edition CTAs
- `<ImagePreview />` displays a preview of an image and of its result without background
- `<NewFolderModal />`  and `<MoveImageModal />` are form as modal
- all these components are glued together in `<App />` 

## Caveats
- Missing CSS 💅: I prioritized functional over appearance. The barebone of UI (sidebar, tabs..) is there though.
- `App.tsx` might need a bit of refactoring to make it clearer (ex: move `uploadImageToServer` to a separate file..)
- form `Modals` could use a bit of factorization which would make code cleaner and help with consistency
- `useImageLibrary` would deserve to be implemented with `useReducer` because the `useImageLibrary` complexity grows and some functions (like `addResultToImage`) depends on previous state which is not ideal with `useState`. See https://github.com/alexstrat/photoroom-test-assignment/pull/1#issuecomment-991065763 =>  done in 2dbdea9

## Test instructions
- upload in a image 
- => it should appear in "Untitled Folder"
- click on "Without Background"
- => you should see the image without background
- upload an other image 
- => it should appear in "Untitled Folder"
- go back to previous image by clicking its name
- create a folder
- => created
- upload in a image in the new folder
- => it appears in the new folder
- move an image from one folder to an other
- => it works
